dftimewolf/lib/exporters/df_to_filesystem.py

```diff
@@ -0,0 +1,176 @@
+    """Set up the module.
+
+    Args:
+      output_formats: Comma separated formats to export. Supported values are:
+          csv, jsonl, markdown. If not specified, 'jsonl' is used.
+      output_directory: Where to write the output. The directory is created if
+          it doesn't already exist.
+    """
+    if output_formats:
+      self._formats = [
+        s.strip().lower() for s in output_formats.split(',') if s]
+      self._formats = list(filter(None, self._formats))
+
+      invalid_formats = []
+      for f in self._formats:
+        if f not in _VALID_OUTPUTS:
+          invalid_formats.append(f)
+      if invalid_formats:
+        self.ModuleError(
+          f'Invalid format(s) specified: {", ".join(invalid_formats)}',
+          critical=True)
+    else:
+      self._formats = [_JSONL]
+
+    self._output_dir = self._VerifyOrCreateOutputDirectory(output_directory)
+
+  def Process(self) -> None:
+    """Perform the exports."""
+    to_export = self.state.GetContainers(containers.DataFrame)
+
+    for df in to_export:
+      self._ExportSingleDataFrame(df)
+
+  def _VerifyOrCreateOutputDirectory(self, directory: str | None) -> str:
+    """Checks for or creates an output directory.
+
+    If the output directory is not specified, then a temp directory will
+    be created. Otherwise, the provided value is checked for existence, and if
+    it doesn't exist, is created.
+
+    Args:
+      directory: The directory path.
+    """
+    if not directory:
+      return tempfile.mkdtemp()
+
+    if os.path.exists(directory):
+      if not os.path.isdir(directory):
+        self.ModuleError(
+          f'Output path {directory} already exists and is not a directory.',
+          critical=True)
+      return directory
+    os.mkdir(directory)
+    return directory
+
+  def _ExportSingleDataFrame(self, container: containers.DataFrame) -> None:
+    """Export a single Dataframe container.
+
+    Args:
+      df_cont: The dataframe container to export.
+    """
+    if _JSONL in self._formats:
+      self._ExportDataFrameJSONL(container)
+    if _CSV in self._formats:
+      self._ExportDataFrameCSV(container)
+    if _MARKDOWN in self._formats:
+      self._ExportDataFrameMarkdown(container)
+
+  def _ExportDataFrameJSONL(self, container: containers.DataFrame) -> None:
+    """Exports a single dataframe container to a jsonl file."""
+    output_path = os.path.join(
+      self._output_dir, f'{_ConvertToValidFilename(container.name)}.{_JSONL}')
+    self.logger.debug(f'Exporting {container.name} to {output_path}')
+
+    with open(output_path, 'w') as f:
+      container.data_frame.to_json(
+        f, orient='records', lines=True, default_handler=str)
+
+    self.state.StoreContainer(container=containers.File(
+        name=os.path.basename(output_path),
+        path=output_path,
+        description=container.description))
+
+    self.logger.debug(f'Export of {container.name} to {output_path} complete')
```

Comment:
This function is nearly identical to `_ExportDataFrameCSV` and `_ExportDataFrameMarkdown`, except for the pandas `to_` function being called. Can we factor that code into `_ExportSingleDataFrame` and to the switch case around the `to_` function?

Reply:
Yeah, good call. Updated, PTAL.